File: examples/python/kem_demo.py
#!/usr/bin/env python3
"""
QUAC 100 ML-KEM (Kyber) Key Exchange Demo

Demonstrates complete key encapsulation mechanism workflow:
- Key generation
- Encapsulation (sender side)
- Decapsulation (receiver side)
- Shared secret verification

Usage:
    python kem_demo.py [512|768|1024]

Copyright 2025 Dyber, Inc. All Rights Reserved.
"""

import sys
import argparse
import secrets
from dataclasses import dataclass
from typing import Tuple, Optional

try:
    from quac100 import QUAC, QUACError, Algorithm
    HAVE_QUAC = True
except ImportError:
    HAVE_QUAC = False
    print("Note: quac100 package not installed. Using simulation mode.\n")


@dataclass
class KEMParams:
    """KEM algorithm parameters"""
    algorithm: str
    name: str
    pk_size: int
    sk_size: int
    ct_size: int
    ss_size: int


# Algorithm parameters
KEM_PARAMS = {
    512: KEMParams("ML_KEM_512", "ML-KEM-512", 800, 1632, 768, 32),
    768: KEMParams("ML_KEM_768", "ML-KEM-768", 1184, 2400, 1088, 32),
    1024: KEMParams("ML_KEM_1024", "ML-KEM-1024", 1568, 3168, 1568, 32),
}


class SimulatedKEM:
    """Simulated KEM for demonstration without hardware"""
    
    def __init__(self, params: KEMParams):
        self.params = params
    
    def keygen(self) -> Tuple[bytes, bytes]:
        """Generate keypair"""
        pk = secrets.token_bytes(self.params.pk_size)
        sk = secrets.token_bytes(self.params.sk_size)
        return pk, sk
    
    def encaps(self, pk: bytes) -> Tuple[bytes, bytes]:
        """Encapsulate to public key"""
        ct = secrets.token_bytes(self.params.ct_size)
        ss = secrets.token_bytes(self.params.ss_size)
        # Store for simulation
        self._last_ss = ss
        return ct, ss
    
    def decaps(self, ct: bytes, sk: bytes) -> bytes:
        """Decapsulate with secret key"""
        # In simulation, return the same shared secret
        return getattr(self, '_last_ss', secrets.token_bytes(self.params.ss_size))


def print_hex(label: str, data: bytes, max_bytes: int = 48):
    """Print data as hex with label"""
    hex_str = data[:max_bytes].hex()
    suffix = "..." if len(data) > max_bytes else ""
    print(f"{label} ({len(data)} bytes):")
    print(f"  {hex_str}{suffix}")


def run_demo(level: int, use_simulator: bool = True):
    """Run the KEM demonstration"""
    
    params = KEM_PARAMS.get(level)
    if not params:
        print(f"Error: Invalid level {level}. Use 512, 768, or 1024.")
        return 1
    
    print("=" * 64)
    print(f"  QUAC 100 ML-KEM Key Exchange Demo")
    print(f"  Algorithm: {params.name} (FIPS 203)")
    print("=" * 64)
    print()
    
    # Initialize
    if HAVE_QUAC and not use_simulator:
        try:
            quac = QUAC()
            kem = quac
            print("Using QUAC 100 hardware accelerator.\n")
        except QUACError:
            kem = SimulatedKEM(params)
            print("Hardware not available, using simulator.\n")
    else:
        kem = SimulatedKEM(params)
        print("Using software simulator.\n")
    
    # =========================================================================
    # Step 1: Key Generation (Receiver - Alice)
    # =========================================================================
    print("Step 1: Key Generation (Receiver - Alice)")
    print("-" * 45)
    print("Alice generates a keypair to receive encrypted messages.\n")
    
    if isinstance(kem, SimulatedKEM):
        alice_pk, alice_sk = kem.keygen()
    else:
        alice_pk, alice_sk = kem.kem_keygen(getattr(Algorithm, params.algorithm))
    
    print_hex("Alice's Public Key", alice_pk)
    print(f"Alice's Secret Key: {len(alice_sk)} bytes (kept private)\n")
    print("Alice sends her public key to Bob...\n")
    
    # =========================================================================
    # Step 2: Encapsulation (Sender - Bob)
    # =========================================================================
    print("Step 2: Encapsulation (Sender - Bob)")
    print("-" * 40)
    print("Bob uses Alice's public key to create:")
    print("  - A ciphertext (to send to Alice)")
    print("  - A shared secret (kept by Bob)\n")
    
    if isinstance(kem, SimulatedKEM):
        ciphertext, ss_bob = kem.encaps(alice_pk)
    else:
        ciphertext, ss_bob = kem.kem_encaps(
            getattr(Algorithm, params.algorithm), alice_pk
        )
    
    print_hex("Ciphertext", ciphertext)
    print_hex("Bob's Shared Secret", ss_bob)
    print("\nBob sends the ciphertext to Alice...\n")
    
    # =========================================================================
    # Step 3: Decapsulation (Receiver - Alice)
    # =========================================================================
    print("Step 3: Decapsulation (Receiver - Alice)")
    print("-" * 42)
    print("Alice uses her secret key to recover the shared secret.\n")
    
    if isinstance(kem, SimulatedKEM):
        ss_alice = kem.decaps(ciphertext, alice_sk)
    else:
        ss_alice = kem.kem_decaps(
            getattr(Algorithm, params.algorithm), ciphertext, alice_sk
        )
    
    print_hex("Alice's Shared Secret", ss_alice)
    print()
    
    # =========================================================================
    # Step 4: Verification
    # =========================================================================
    print("Step 4: Verification")
    print("-" * 22)
    
    if ss_bob == ss_alice:
        print("✓ SUCCESS! Both parties have the same shared secret.")
        print("  This secret can now be used as a symmetric encryption key.\n")
    else:
        print("✗ FAILURE! Shared secrets do not match.")
        return 1
    
    # =========================================================================
    # Summary
    # =========================================================================
    print("=" * 64)
    print("  Key Exchange Complete")
    print("=" * 64)
    print(f"Algorithm:      {params.name}")
    print(f"Public Key:     {len(alice_pk)} bytes")
    print(f"Secret Key:     {len(alice_sk)} bytes")
    print(f"Ciphertext:     {len(ciphertext)} bytes")
    print(f"Shared Secret:  {len(ss_bob)} bytes (256 bits)")
    print()
    print("This shared secret provides:")
    print("  • Post-quantum security against Shor's algorithm")
    print("  • IND-CCA2 security (chosen ciphertext attack resistance)")
    print("  • Perfect forward secrecy when used with ephemeral keys")
    print("=" * 64)
    
    return 0


def main():
    parser = argparse.ArgumentParser(
        description="QUAC 100 ML-KEM Key Exchange Demo"
    )
    parser.add_argument(
        "level",
        nargs="?",
        type=int,
        default=768,
        choices=[512, 768, 1024],
        help="Security level: 512, 768, or 1024 (default: 768)"
    )
    parser.add_argument(
        "-s", "--simulator",
        action="store_true",
        help="Force use of software simulator"
    )
    
    args = parser.parse_args()
    return run_demo(args.level, args.simulator)


if __name__ == "__main__":
    sys.exit(main())

File: examples/rust/kem_demo.rs
//! QUAC 100 ML-KEM (Kyber) Key Exchange Demo (Rust)
//!
//! Demonstrates complete key encapsulation mechanism workflow:
//! - Key generation
//! - Encapsulation (sender side)
//! - Decapsulation (receiver side)
//! - Shared secret verification
//!
//! Run: cargo run --example kem_demo -- [512|768|1024]
//!
//! Copyright 2025 Dyber, Inc. All Rights Reserved.

use std::env;

// Simulated types (would be imported from quac100 crate)
#[derive(Clone, Copy)]
enum Algorithm {
    MlKem512,
    MlKem768,
    MlKem1024,
}

struct KemParams {
    name: &'static str,
    pk_size: usize,
    sk_size: usize,
    ct_size: usize,
    ss_size: usize,
}

impl KemParams {
    fn get(level: u32) -> Self {
        match level {
            512 => KemParams {
                name: "ML-KEM-512",
                pk_size: 800,
                sk_size: 1632,
                ct_size: 768,
                ss_size: 32,
            },
            768 => KemParams {
                name: "ML-KEM-768",
                pk_size: 1184,
                sk_size: 2400,
                ct_size: 1088,
                ss_size: 32,
            },
            1024 => KemParams {
                name: "ML-KEM-1024",
                pk_size: 1568,
                sk_size: 3168,
                ct_size: 1568,
                ss_size: 32,
            },
            _ => Self::get(768),
        }
    }
}

/// Simulated KEM for demonstration
struct SimulatedKem {
    params: KemParams,
    last_ss: Option<Vec<u8>>,
}

impl SimulatedKem {
    fn new(params: KemParams) -> Self {
        Self { params, last_ss: None }
    }

    fn keygen(&self) -> (Vec<u8>, Vec<u8>) {
        use std::collections::hash_map::DefaultHasher;
        use std::hash::{Hash, Hasher};
        use std::time::{SystemTime, UNIX_EPOCH};

        let seed = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos();

        let mut hasher = DefaultHasher::new();
        seed.hash(&mut hasher);

        let pk: Vec<u8> = (0..self.params.pk_size)
            .map(|i| {
                let mut h = DefaultHasher::new();
                (seed + i as u128).hash(&mut h);
                (h.finish() & 0xFF) as u8
            })
            .collect();

        let sk: Vec<u8> = (0..self.params.sk_size)
            .map(|i| {
                let mut h = DefaultHasher::new();
                (seed + i as u128 + 1000000).hash(&mut h);
                (h.finish() & 0xFF) as u8
            })
            .collect();

        (pk, sk)
    }

    fn encaps(&mut self, _pk: &[u8]) -> (Vec<u8>, Vec<u8>) {
        use std::collections::hash_map::DefaultHasher;
        use std::hash::{Hash, Hasher};
        use std::time::{SystemTime, UNIX_EPOCH};

        let seed = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos();

        let ct: Vec<u8> = (0..self.params.ct_size)
            .map(|i| {
                let mut h = DefaultHasher::new();
                (seed + i as u128).hash(&mut h);
                (h.finish() & 0xFF) as u8
            })
            .collect();

        let ss: Vec<u8> = (0..self.params.ss_size)
            .map(|i| {
                let mut h = DefaultHasher::new();
                (seed + i as u128 + 2000000).hash(&mut h);
                (h.finish() & 0xFF) as u8
            })
            .collect();

        self.last_ss = Some(ss.clone());
        (ct, ss)
    }

    fn decaps(&self, _ct: &[u8], _sk: &[u8]) -> Vec<u8> {
        // Return stored secret for simulation
        self.last_ss.clone().unwrap_or_else(|| vec![0u8; self.params.ss_size])
    }
}

fn print_hex(label: &str, data: &[u8], max_bytes: usize) {
    let show = data.len().min(max_bytes);
    let hex: String = data[..show].iter().map(|b| format!("{:02x}", b)).collect();
    let suffix = if data.len() > max_bytes { "..." } else { "" };
    println!("{} ({} bytes):", label, data.len());
    println!("  {}{}", hex, suffix);
}

fn main() {
    // Parse command line
    let args: Vec<String> = env::args().collect();
    let level: u32 = args.get(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(768);

    if ![512, 768, 1024].contains(&level) {
        eprintln!("Error: Invalid level. Use 512, 768, or 1024.");
        std::process::exit(1);
    }

    let params = KemParams::get(level);

    println!("================================================================");
    println!("  QUAC 100 ML-KEM Key Exchange Demo (Rust)");
    println!("  Algorithm: {} (FIPS 203)", params.name);
    println!("================================================================");
    println!();

    // Initialize
    let mut kem = SimulatedKem::new(params);
    println!("Using software simulator.");
    println!();

    // =========================================================================
    // Step 1: Key Generation (Receiver - Alice)
    // =========================================================================
    println!("Step 1: Key Generation (Receiver - Alice)");
    println!("------------------------------------------");
    println!("Alice generates a keypair to receive encrypted messages.");
    println!();

    let (alice_pk, alice_sk) = kem.keygen();

    print_hex("Alice's Public Key", &alice_pk, 48);
    println!("Alice's Secret Key: {} bytes (kept private)", alice_sk.len());
    println!();
    println!("Alice sends her public key to Bob...");
    println!();

    // =========================================================================
    // Step 2: Encapsulation (Sender - Bob)
    // =========================================================================
    println!("Step 2: Encapsulation (Sender - Bob)");
    println!("-------------------------------------");
    println!("Bob uses Alice's public key to create:");
    println!("  - A ciphertext (to send to Alice)");
    println!("  - A shared secret (kept by Bob)");
    println!();

    let (ciphertext, ss_bob) = kem.encaps(&alice_pk);

    print_hex("Ciphertext", &ciphertext, 48);
    print_hex("Bob's Shared Secret", &ss_bob, 32);
    println!();
    println!("Bob sends the ciphertext to Alice...");
    println!();

    // =========================================================================
    // Step 3: Decapsulation (Receiver - Alice)
    // =========================================================================
    println!("Step 3: Decapsulation (Receiver - Alice)");
    println!("-----------------------------------------");
    println!("Alice uses her secret key to recover the shared secret.");
    println!();

    let ss_alice = kem.decaps(&ciphertext, &alice_sk);

    print_hex("Alice's Shared Secret", &ss_alice, 32);
    println!();

    // =========================================================================
    // Step 4: Verification
    // =========================================================================
    println!("Step 4: Verification");
    println!("--------------------");

    if ss_bob == ss_alice {
        println!("✓ SUCCESS! Both parties have the same shared secret.");
        println!("  This secret can now be used as a symmetric encryption key.");
        println!();
    } else {
        println!("✗ FAILURE! Shared secrets do not match.");
        std::process::exit(1);
    }

    // =========================================================================
    // Summary
    // =========================================================================
    println!("================================================================");
    println!("  Key Exchange Complete");
    println!("================================================================");
    println!("Algorithm:      {}", params.name);
    println!("Public Key:     {} bytes", alice_pk.len());
    println!("Secret Key:     {} bytes", alice_sk.len());
    println!("Ciphertext:     {} bytes", ciphertext.len());
    println!("Shared Secret:  {} bytes (256 bits)", ss_bob.len());
    println!();
    println!("This shared secret provides:");
    println!("  • Post-quantum security against Shor's algorithm");
    println!("  • IND-CCA2 security (chosen ciphertext attack resistance)");
    println!("  • Perfect forward secrecy when used with ephemeral keys");
    println!("================================================================");

    // Note: In real code, use zeroize crate to securely clear secrets
    // alice_sk.zeroize();
    // ss_bob.zeroize();
    // ss_alice.zeroize();
}